dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/AbstractImageRegistry.java

```diff
@@ -419,18 +425,91 @@ private void updateFiles(Tool tool, Tag tag, final FileDAO fileDAO, SourceCodeRe
             tag.addSourceFile(file);
         }
 
-        // need to go through all files for the booleans
-        boolean hasDockerfile = tag.getSourceFiles().stream().anyMatch(file -> file.getType() == SourceFile.FileType.DOCKERFILE);
+        // Update the tag with validation information
+        tag = validateTagDockerfile(tag, tool.isPrivateAccess());
+        tag = validateTagDescriptorType(tag, SourceFile.FileType.DOCKSTORE_CWL, tag.getCwlPath());
+        tag = validateTagDescriptorType(tag, SourceFile.FileType.DOCKSTORE_WDL, tag.getWdlPath());
+
+        boolean isValidVersion = isValidVersion(tag);
+        tag.setValid(isValidVersion);
+    }
+
+    /**
+     * Checks if the given tag is valid given its version validations.
+     * TODO: Duplicate in HostedToolResource.java
+     * @param tag Tag to check validation
+     * @return True if valid tag, false otherwise
+     */
+    private boolean isValidVersion(Tag tag) {
+        SortedSet<VersionValidation> versionValidations = tag.getValidations();
+        boolean validDockerfile = isVersionTypeValidated(versionValidations, SourceFile.FileType.DOCKERFILE);
+        boolean validCwl = isVersionTypeValidated(versionValidations, SourceFile.FileType.DOCKSTORE_CWL);
+        boolean validWdl = isVersionTypeValidated(versionValidations, SourceFile.FileType.DOCKSTORE_WDL);
+        boolean validCwlTestParameters = isVersionTypeValidated(versionValidations, SourceFile.FileType.CWL_TEST_JSON);
+        boolean validWdlTestParameters = isVersionTypeValidated(versionValidations, SourceFile.FileType.WDL_TEST_JSON);
+
         boolean hasCwl = tag.getSourceFiles().stream().anyMatch(file -> file.getType() == SourceFile.FileType.DOCKSTORE_CWL);
         boolean hasWdl = tag.getSourceFiles().stream().anyMatch(file -> file.getType() == SourceFile.FileType.DOCKSTORE_WDL);
 
+        return validDockerfile && ((hasCwl && validCwl && validCwlTestParameters) || (hasWdl && validWdl && validWdlTestParameters));
+    }
 
+    /**
+     * Finds the first occurrence of a specific sourcefile type in a set of validations and returns whether or not it is valid
+     * @param versionValidations Set of version validations
+     * @param fileType File Type to look for
+     * @return True if exists and valid, false otherwise
+     */
+    private boolean isVersionTypeValidated(SortedSet<VersionValidation> versionValidations, SourceFile.FileType fileType) {
```

Comment:
This seems oddly non-deterministic. 
Given a filetype with an invalid and valid validations, this will return different results based on order in the set?